elastic/kibana
# 130186 update firstlastseen query to be used by all details pages

## Summary
[An SDH ticket was recently generated ](https://github.com/elastic/sdh-security-team/issues/342) due to the Network details (IP Overview page) loading very slowly. Specifically the concern was around the top pane:
![image](https://user-images.githubusercontent.com/28942857/171928781-7e13af61-e121-431d-bd6f-bc304958646d.png)

Upon further investigation, it was noted that a part of the query-  (first_seen / last_seen), was unbounded in time) and also an expensive query to run. 


A few months back, FrankHassanabad [did some digging into this type of query ](https://github.com/elastic/kibana/pull/91790) and found some things worth noting. Here's an excerpt:

> using aggregations and max/min [...] is slower in orders of magnitude vs. using sorting + size: 1 + track_total_hits: false.
> 
> **Adding track_total_hits: false to aggregations does not appear to make things faster, nor make them slower. It seems to have no effect during testing and experimentation. 
> 
> However, adding and using track_total_hits: false to regular queries does speed things up considerably and using it with size: 1 and sort looks to be very promising and fast. So, it looks like where and when we can replace an aggregation query that has a max or min with a regular query/filter + sort + track_total_hits: false we are going to be 👍

As such, this PR aims to make this type of query accessible to all views needing last/first seen by shifting its existence from only the hosts page where Frank implemented it to common, and using it for the Network and User detail pages.

This PR is broken into 4 main phases, and it might be helpful to consider it commit by commit.

The first phase is moving the files existing in the hosts folder related to the first_last_seen the ( component and container ) to a common directory and updating types to be more generic.
 - a move from first_last_seen_host -> first_last_seen
 - and updating props from `hostName` to `field` (which field the search will be for in ES format like `host.name` or `user.name`, and a value such as `KristofsMacBook`

The second commit updates usage of this component to not just be host overview, but also user and network overview.
 
The third commit pertains to the server and updating types and usage to allow for this updated capability.

Finally all following commits focus on cleaning up types and fixing / adding tests.

## QA Testing:

The queries made to load the details table for the users, hosts, and network details pages, all include a section:
`First seen`, and `Last seen`. These queries are now separated from the other data seen on this page because they can take a very long time to load causing slow UI problems.

To test, everything should work as before. The main UI change is those specific queries may take longer to load, but you will see a loading spinner next to them in that case, just as it was on the Host Details page:

![quantify_perf](https://user-images.githubusercontent.com/1151048/108309335-99773600-716e-11eb-8dae-c289947b188c.gif)


An unfortunate side effect of this work is the query inspector will no longer show the part of the query related to first or last seen. This is akin to how it is already on the host details page, and for now the pros in efficiency outweighs this con.
